Cython/Compiler/Nodes.py

```diff
@@ -1226,11 +1227,15 @@ class TemplatedTypeNode(CBaseTypeNode):
     name = None
 
     def _analyse_template_types(self, env, base_type):
-        require_python_types = base_type.python_type_constructor_name in (
-            'typing.Optional',
-            'dataclasses.ClassVar',
-        )
-        in_c_type_context = env.in_c_type_context and not require_python_types
+
+        required_types = {
+            'typing.Optional': lambda t: t.is_pyobject or t.is_buffer or t.is_memoryviewslice,
+            'dataclasses.ClassVar': lambda t: t.is_pyobject,
+        }
+
+        check_require_types = required_types.get(base_type.python_type_constructor_name, None)
+
+        in_c_type_context = env.in_c_type_context and check_require_types is not None
```

Comment:
Shouldn't it be `is None` instead of `is not None`?
```suggestion
        in_c_type_context = env.in_c_type_context and check_require_types is None
```